File: GUILib/src/space/kiibou/gui/VerticalList.java
package space.kiibou.gui;

import space.kiibou.GApplet;

public class VerticalList extends GraphicsElement {
    public VerticalList(GApplet app, int x, int y, int scale) {
        super(app, x, y, 0, 0, scale);
    }

    @Override
    protected void preInitImpl() {

    }

    @Override
    public void initImpl() {
        final int[] height = {0};
        getChildren().forEach(child -> {
            child.moveTo(getX(), getY() + height[0]);
            height[0] += child.getHeight();
        });

        int width = 0;
        if (getChildren().size() > 0) {
            width = getChildren().stream().mapToInt(GraphicsElement::getWidth).max().orElse(0);
            for (GraphicsElement child : getChildren()) child.setWidth(width);
        }
        resize(width, height[0]);
    }

    @Override
    protected void postInitImpl() {

    }

    @Override
    protected void drawImpl() {

    }

    @Override
    public void addChild(GraphicsElement element) {
        BorderBox borderBox = new BorderBox(getApp(), getScale());
        borderBox.addChild(element);
        super.addChild(borderBox);
    }

    @Override
    public BorderBox getChild(int index) {
        return (BorderBox) super.getChild(index);
    }

}

File: GUILib/src/space/kiibou/gui/VerticalList.kt
package space.kiibou.gui

import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.value.ObservableNumberValue
import space.kiibou.GApplet

class VerticalList(app: GApplet, x: Int, y: Int, margin: Int = 0, scale: Int) : GraphicsElement(app, x, y, 0, 0, scale) {

    private val marginProp = scaleProp.multiply(margin)

    init {
        childrenProperty.addListener { _, _, list ->
            var h: ObservableNumberValue = SimpleIntegerProperty(0)
            list.forEach {
                it.xProp.bind(xProp)
                it.yProp.bind(yProp.add(h))
                h = it.heightProp.add(h).add(marginProp)
            }
            heightProp.bind(Bindings.subtract(h, marginProp))

            var w: ObservableNumberValue = SimpleIntegerProperty(0)
            list.forEach { w = Bindings.max(it.widthProp, w) }
            widthProp.bind(w)
        }
    }

    override fun preInitImpl() {}
    public override fun initImpl() {}
    override fun postInitImpl() {}
    override fun drawImpl() {}

}